### PR TITLE
Separate Shaders Into a Subdirectory and Cache Them.

### PR DESCRIPTION
**Pull request by Koranir**
## Feature Details
This PR aims to separate shaders from being defined inside the .cpp files, and instead in a separate source folder called "shaders". This adds some delay while initialising due to needing to read the code from an outside file instead of being built-in to the executable, so I added capability to cache the shader. This also has the added benefit of decreasing the game's startup time by about a second after the initial cache.

Due to the shaders being separate from the .cpp files, it becomes much easier to view and edit them, and the game doesn't need to be recompiled whenever a shader is changed. (This was the main reason I made this, as I am looking into implementing post processing based on the current player system.)

## Usage Examples
before in `FogShader.cpp`:
```
        static const char *vertexCode =
		"// vertex fog shader\n"
		"uniform vec2 corner;\n"
		"uniform vec2 dimensions;\n"
		"in vec2 vert;\n"
		"out vec2 fragTexCoord;\n"
		"void main() {\n"
		"  gl_Position = vec4(corner + vert * dimensions, 0, 1);\n"
		"  fragTexCoord = vert;\n"
		"}\n";
	static const char *fragmentCode =
		"// fragment fog shader\n"
#ifdef ES_GLES
		"precision mediump sampler2D;\n"
#endif
		"precision mediump float;\n"
		"uniform sampler2D tex;\n"
		"in vec2 fragTexCoord;\n"
		"out vec4 finalColor;\n"
		"void main() {\n"
		"  finalColor = vec4(0, 0, 0, texture(tex, fragTexCoord).r);\n"
		"}\n";
	// Compile the shader and store indices to its variables.
	shader = Shader(vertexCode, fragmentCode);
```
after in `FogShader.cpp`:
```
	shader = Shader("fog", true);
```
The `Shader()` function was changed to accept a string and two bools instead of two char arrays. The string is the name of the shader and is used as the path to the folder where the shader code is. The first bool tells the function that it should look in the "inbuilt" subdirectory of "shaders" before the name of the shader is added, and the second is to tell the function whether to use shader swizzle.

The actual shader code was added into files `source/shaders/fog/FragmentCode.txt` &
`source/shaders/fog/VertexCode.txt` and is read dynamically.

A disadvantage to this approach is not being able to use preprocessor definitions to filter out features for different versions, e.g. GLES letting you define precisions for vectors, but those definition causing compiler errors on older systems. I could add more shader files that have those different features in them and use the preproccessor definitions in the .cpp files to switch between them, but I'd like a more permanent solution.

in `shaders/inbuilt/fog/FragmentCode.txt`:
```
// fragment fog shader
// precision mediump sampler2D;
precision mediump float;
uniform sampler2D tex;

in vec2 fragTexCoord;
out vec4 finalColor;

void main()
{
  finalColor = vec4(0, 0, 0, texture(tex, fragTexCoord).r);
}
```
See, easy.

## Testing Done
I've only been able to do very limited testing due to only testing on one machine that supports all the features used. I need some others who have older OpenGL versions to test this out.

## TODO:

- [x] Check compatibility with older OpenGL versions
- [x] Change indent type in the Shader code to use tabs, not whitespaces

## Performance Impact
Shader loading time before PR was 1.33 seconds.
Shader loading time with the PR but not cached was 1.44 seconds.
Shader loading time with PR and Shaders cached was 0.39 seconds.

### CODE REVIEW

**Comment by quyykk** [comment]
> This adds some delay due to needing to read from an outside file instead of being built-in to the executable, so I added capability to cache the shader

Why not have the shader files be read once at startup and compiled?

> A disadvantage to this approach is not being able to use preprocessor definitions to filter out features for different version

I think it's possible to use preprocessor checks inside the GLSL files too. So we can leave the shader code as-is but insert a `#define ES_GLES` when compiling the shader if we're building for GLES.

**Comment by Amazinite** [comment]
#2442
> A more proper solution to this in the future might be to somehow cache the output of OutlineShader so that it only needs to be computed once, then just show the cached outline from there on out, but I don't have the knowledge for that.

Is it finally happening? 👀

**Comment by Koranir** [comment]
> > This adds some delay due to needing to read from an outside file instead of being built-in to the executable, so I added capability to cache the shader
> 
> Why not have the shader files be read once at startup and compiled?

I think I explained this wrong in the description, the files are read once at startup, compiled and cached if there is no cached shader present, otherwise the cache will be read to the shader program.

> > A disadvantage to this approach is not being able to use preprocessor definitions to filter out features for different version
> 
> I think it's possible to use preprocessor checks inside the GLSL files too. So we can leave the shader code as-is but insert a `#define ES_GLES` when compiling the shader if we're building for GLES.

I'll look into this.


> #2442
> 
> > A more proper solution to this in the future might be to somehow cache the output of OutlineShader so that it only needs to be computed once, then just show the cached outline from there on out, but I don't have the knowledge for that.
> 
> Is it finally happening? 👀

This isn't a solution for that issue in particular, but I will be looking into that issue now.

**Comment by quyykk** [comment]
> Blaming Quyykk if this fails

~~you forgot to do this:~~ You can pass in an additional source text for a given shader: [glShaderSource](https://registry.khronos.org/OpenGL-Refpages/gl4/html/glShaderSource.xhtml). So there you would pass `#define ES_GLES` when appropriate.

File: source/Shader.cpp [hunk 1]
```diff
@@ -151,3 +306,15 @@ GLuint Shader::Compile(const char *str, GLenum type)
...
 }
+
+
+
+string Shader::ShaderPath(string shader, bool isFragment, bool isInBuilt, bool useShaderSwizzle)
+{
+	string path = Files::Shaders();
+	path += isInBuilt ? "inbuilt/" : "";
+	path += shader;
+	path += useShaderSwizzle ? "/swizzle/" : "/";
```
Comment (quyykk) [inline source/Shader.cpp:317]: You're appending swizzle to the shader paths, this only works for the sprite shader and for every other the path will be incorrect.

**Review by quyykk — COMMENTED** [review]
I think we should use the standardized (more or less) file extensions for the shaders: `.vert` and `.frag`. That way we also avoid having a lot of directories and files with the same exact name.

**Comment by RisingLeaf** [comment]
@Koranir you have no interest in this anymore?

**Pull request closed without merging**